### PR TITLE
Introduce ManufacturerIdInterface for ValueObjects

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Introduce `ManufacturerIdInterface ` interface to be used as a transverse type instead of `ManufacturerId` which is restricted to more accurate cases
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | ~
| How to test?  | Travis tests green

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20828)

### CODE REVIEW

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

File: src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+namespace PrestaShop\PrestaShop\Core\Domain\Manufacturer\ValueObject;
+
+/**
+ * Class ManufacturerRelation represents a relation with a Manufacturer, it is not usable as
+ * a stand alone class though, you need to to use one of its children.
+ *
+ * @see ManufacturerId
+ * @see NoManufacturer
+ */
+abstract class ManufacturerRelation
```
Comment (eternoendless) [inline src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php]: I still think this should be called `OptionalManufacturerId`. An ID is not a relation but an identification.

File: src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+{
+    /**
+     * @var int
+     */
+    protected $id;
+
+    /**
+     * @param int $id
+     */
+    public function __construct(int $id)
```
Comment (eternoendless) [inline src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php]: This one should only exist in `ManufacturerId`

File: src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        $this->id = $id;
+    }
+
+    /**
+     * @return int
+     */
+    public function getValue()
```
Comment (eternoendless) [inline src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php]: This one should be abstract

File: src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @see ManufacturerId
+ * @see NoManufacturer
+ */
+abstract class ManufacturerRelation
+{
+    /**
+     * @var int
+     */
+    protected $id;
```
Comment (eternoendless) [inline src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php]: IMO this should be removed

File: src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php (hunk 1 above)
Comment (eternoendless) [reply src/Core/Domain/Manufacturer/ValueObject/ManufacturerRelation.php]: This _could_ become an interface as per my comments below

File: src/Core/Domain/Manufacturer/ValueObject/NoManufacturer.php [hunk 5]
```diff
@@ -47,6 +47,6 @@ class NoManufacturerId extends ManufacturerId
      */
     public function __construct()
     {
-        $this->id = static::NO_MANUFACTURER_ID;
+        parent::__construct(static::NO_MANUFACTURER_ID);
```
Comment (eternoendless) [inline src/Core/Domain/Manufacturer/ValueObject/NoManufacturer.php]: You should implement `getValue()` instead

File: src/Core/Domain/Manufacturer/ValueObject/ManufacturerId.php [hunk 6]
```diff
@@ -46,15 +46,7 @@ class ManufacturerId
...
-        $this->id = $id;
-    }
-
-    /**
-     * @return int
-     */
-    public function getValue()
-    {
-        return $this->id;
+        parent::__construct((int) $id);
```
Comment (zuk3975) [inline src/Core/Domain/Manufacturer/ValueObject/ManufacturerId.php]: you should not cast `int` here I suppose.
VO should not allow invalid data, but it shouldn't be fixing it.

**Review by atomiix — APPROVED** [review]

**Comment by zuk3975** [comment]
soo, we have an interface for a value object now :thinking:

**Comment by matks** [comment]
> soo, we have an interface for a value object now 🤔

I'm puzzle too 😄  but let's give it a chance!

**Review by matthieu-rolland — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
Would be awesome to have also @eternoendless to confirm we are following the right track 😄

**Comment by matks** [comment]
Since @eternoendless is off for 1 week, I believe we can safely merge this PR as it's small and change our course later if we found out an issue

**Pull request merged**